targets/mbedos5/tools/js2c.py

```diff
@@ -0,0 +1,276 @@
+    print("Updated pins.js")
+
+jshint_is_installed = False
+try:
+    with open(os.devnull, "w") as dev_null:
+        if platform.system() == "Windows":
+            subprocess.check_call("where jshint", shell=True, stdout=dev_null)
+        else:
+            subprocess.check_call("which jshint", shell=True, stdout=dev_null)
+        jshint_is_installed = True
+except:
+    print("No JSHint detected, cannot statically check if PinNames you use are valid for your platform.")
+    print("Consider installing JSHint for this feature.")
+
+if jshint_is_installed and platform.system() != "Windows":
+    try:
+        subprocess.check_call("tools/check_pins.sh")
+    except:
+        print("JSHint picked up an undefined variable in your code.")
+        print("Are you using a PinName that doesn't exist for your platform?")
+        raise
+        sys.exit(1)
+
+fout = open(OUT_PATH + 'js_encoded.cpp', 'w')
+fout.write(LICENSE);
+fout.write(HEADER);
+
+pins = list()
+pin_def_re = re.compile("var ([a-zA-Z_0-9]+)\s*=\s*([^;]*);")
+
+def exportOneFile(path, name):
+    # Prepend pins.js onto main.js, don't include pins.js itself.
+    code = ""
+    if name == "pins":
+        global pins, pin_def_re
+        fin = open(path, "r")
+        for line in fin.readlines():
+            m = pin_def_re.search(line.strip())
+            if m:
+                pins.append((m.group(1), m.group(2)))
+        fin.close()
+        return
+
+    fout.write('const static char ' + name + '_n[] = "' + name + '";\n')
+    fout.write('const static char ' + name + '_s[] =\n{\n')
+
+    fin = open(path, 'r');
+    code += fin.read() + '\0'
+
+    # minimize code when release mode
+    if buildtype != 'debug':
+        code = removeComments(code)
+        code = removeWhitespaces(code)
+
+    for line in regroup(code, 10):
+        buf = ', '.join(map(lambda ch: format(ord(ch),"#04x"), line))
+        if line[-1] != '\0':
+            buf += ','
+        writeLine(fout, buf, 1)
+    writeLine(fout, '};')
+    writeLine(fout, 'const static int ' + name + '_l = ' + str(len(code)-1) + ';')
+    writeLine(fout, '')
+
+    fin.close();
+
+def exportOneName(name):
+    writeLine(fout, '{ ' + name + '_n, ' + name + '_s, ' + name + '_l },', 1)
+
+files = glob.glob(SRC_PATH + '*.js')
+for path in files:
+    name = extractName(path)
+    exportOneFile(path, name)
+
+
+NATIVE_STRUCT = '''
+struct jsmbed_js_source_t jsmbed_js_code_storage[] =
+{
+'''
+
+fout.write(NATIVE_STRUCT)
+exportOneName('main')
+filenames = sorted(map(extractName, files), key=str.lower)
+for name in filenames:
+    if name != 'main' and name != 'pins':
+        exportOneName(name)
+
+writeLine(fout, '{ 0, 0, 0 }', 1)
+writeLine(fout, '};')
+
+pin_count = len(pins)
+writeLine(fout, 'unsigned int jsmbed_js_magic_string_count = {};'.format(pin_count))
+
+writeLine(fout, 'const char *jsmbed_js_magic_strings[] = {')
+comma = ","
+
+for (idx, (pinname, value)) in enumerate(pins):
```

Comment:
I'm not a big fan of iterating over the same thing more than one times. We could extract the required data in one go. Then write it out when we need it. Also adding the comma manually (basically joining strings) makes me wondering if I can do it better.

In short, maybe something like this can be useful:
```
string_entries = []
length_entries = []
value_entries = []

for (idx, (name, value)) in enaumerate(pins):
    string_entries.append('"{}"'.format(name))
    length_entries.append('{}'.format(len(name)))
    value_entries.append('{}'.format(value))


writeLine(fout, ',\n    '.join(string_entries), 1)
...
writeLine(fout, ',\n    '.join(length_entries), 1)
...
writeLine(fout, ',\n    '.join(value_entries), 1)
```